gh-125843: indicate which C function caused a `curses.error`

picnixz opened the pull request:
<!-- gh-issue-number: gh-125843 -->
* Issue: gh-125843
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--125844.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

picnixz committed:
improve `curses.error` object

picnixz committed:
update error messages

picnixz committed:
cosmetic changes!

picnixz committed:
blurb

picnixz committed:
update docs

picnixz committed:
Merge remote-tracking branch 'upstream/main' into x-stale/feat/curses/error-type-125843

picnixz committed:
reduce diff

picnixz committed:
reduce diff even futher

picnixz committed:
simplify logic

picnixz committed:
simplify logic x2

picnixz committed:
reduce diff again!

picnixz committed:
simplify logic x3

picnixz committed:
lint

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -287,44 +366,52 @@ _PyCursesStatefulCheckFunction(PyObject *module, int called, const char *funcnam
+}
+
+static PyObject *PyCursesCheckERR_ForWin_From(PyCursesWindowObject *, int,
+                                              const char *, const char *);
```
```suggestion
static PyObject *
PyCursesCheckERR_ForWin_From(PyCursesWindowObject *, int,
                             const char *, const char *);
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -1064,27 +1172,37 @@ _curses_window_addnstr_impl(PyCursesWindowObject *self, int group_left_1,
+            curses_funcname = "mvwaddnstr";
+        }
+        else {
+            rtn = waddnstr(self->win, str, n);
```
```suggestion
            rtn = mvwaddnstr(self->win,y,x,str,n);
            curses_funcname = "mvwaddnstr";
        }
        else {
            rtn = waddnstr(self->win,str,n);
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -1378,13 +1501,15 @@ PyCursesWindow_ChgAt(PyObject *op, PyObject *args)
         rtn = mvwchgat(self->win,y,x,num,attr,color,NULL);
-        touchline(self->win,y,1);
+        curses_funcname = "mvwchgat";
+        (void)touchline(self->win,y,1);
```
```suggestion
        (void)touchline(self->win,y,1);
        curses_funcname = "mvwchgat";
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -1378,13 +1501,15 @@ PyCursesWindow_ChgAt(PyObject *op, PyObject *args)
         rtn = wchgat(self->win,num,attr,color,NULL);
-        touchline(self->win,y,1);
+        curses_funcname = "wchgat";
+        (void)touchline(self->win,y,1);
```
```suggestion
        (void)touchline(self->win,y,1);
        curses_funcname = "wchgat";
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -1997,27 +2138,37 @@ _curses_window_insstr_impl(PyCursesWindowObject *self, int group_left_1,
+        }
+        else {
+            rtn = wins_wstr(self->win, wstr);
+            curses_funcname = "wins_wstr";
```
```suggestion
            rtn = mvwins_wstr(self->win,y,x,wstr);
            curses_funcname = "mvwins_wstr";
        }
        else {
            rtn = wins_wstr(self->win,wstr);
            curses_funcname = "wins_wstr";
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -1997,27 +2138,37 @@ _curses_window_insstr_impl(PyCursesWindowObject *self, int group_left_1,
+        else {
+            rtn = winsstr(self->win, text);
+            curses_funcname = "winsstr";
+        }
```
```suggestion
        const char *str = PyBytes_AS_STRING(bytesobj);
        simple_funcname = "insstr";
        if (use_xy) {
            rtn = mvwinsstr(self->win,y,x,str);
            curses_funcname = "mvwinsstr";
        }
        else {
            rtn = winsstr(self->win,str);
            curses_funcname = "winsstr";
        }
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -2082,27 +2233,37 @@ _curses_window_insnstr_impl(PyCursesWindowObject *self, int group_left_1,
+        else {
+            rtn = wins_nwstr(self->win, wstr, n);
+            curses_funcname = "wins_nwstr";
+        }
```
```suggestion
        if (use_xy) {
            rtn = mvwins_nwstr(self->win,y,x,wstr,n);
            curses_funcname = "mvwins_nwstr";
        }
        else {
            rtn = wins_nwstr(self->win,wstr,n);
            curses_funcname = "wins_nwstr";
        }
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -2082,27 +2233,37 @@ _curses_window_insnstr_impl(PyCursesWindowObject *self, int group_left_1,
+        else {
+            rtn = winsnstr(self->win, text, n);
+            curses_funcname = "winsnstr";
+        }
```
```suggestion
        const char *str = PyBytes_AS_STRING(bytesobj);
        simple_funcname = "insnstr";
        if (use_xy) {
            rtn = mvwinsnstr(self->win,y,x,str,n);
            curses_funcname = "mvwinsnstr";
        }
        else {
            rtn = winsnstr(self->win,str,n);
            curses_funcname = "winsnstr";
        }
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -3563,7 +3756,7 @@ _curses_setupterm_impl(PyObject *module, const char *term, int fd)
 
     if (!curses_setupterm_called && setupterm((char *)term, fd, &err) == ERR) {
-        const char* s = "setupterm: unknown error";
+        const char *s = "setupterm: unknown error";
```
```suggestion
        const char* s = "setupterm: unknown error";
```

picnixz committed:
reduce diff x4

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -892,36 +983,44 @@ _curses_window_addch_impl(PyCursesWindowObject *self, int group_left_1,
-        if (coordinates_group)
-            rtn = mvwadd_wch(self->win,y,x, &wcval);
+        if (coordinates_group) {
+            rtn = mvwadd_wch(self->win, y, x, &wcval);
```
```suggestion
            rtn = mvwadd_wch(self->win,y,x, &wcval);
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -892,36 +983,44 @@ _curses_window_addch_impl(PyCursesWindowObject *self, int group_left_1,
-            rtn = mvwaddch(self->win,y,x, cch | (attr_t) attr);
+        simple_funcname = "addch";
+        if (coordinates_group) {
+            rtn = mvwaddch(self->win, y, x, cch | (attr_t) attr);
```
```suggestion
            rtn = mvwaddch(self->win,y,x, cch | (attr_t) attr);
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -981,27 +1080,37 @@ _curses_window_addstr_impl(PyCursesWindowObject *self, int group_left_1,
-            rtn = waddwstr(self->win,wstr);
+        simple_funcname = "addwstr";
+        if (use_xy) {
+            rtn = mvwaddwstr(self->win, y, x, wstr);
```
```suggestion
            rtn = mvwaddwstr(self->win,y,x,wstr);
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -981,27 +1080,37 @@ _curses_window_addstr_impl(PyCursesWindowObject *self, int group_left_1,
+            curses_funcname = "mvwaddwstr";
+        }
+        else {
+            rtn = waddwstr(self->win, wstr);
```
```suggestion
            rtn = waddwstr(self->win,wstr);
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -981,27 +1080,37 @@ _curses_window_addstr_impl(PyCursesWindowObject *self, int group_left_1,
+        else {
+            rtn = waddstr(self->win, text);
+            curses_funcname = "waddstr";
+        }
```
```suggestion
        const char *str = PyBytes_AS_STRING(bytesobj);
        simple_funcname = "addstr";
        if (use_xy) {
            rtn = mvwaddstr(self->win,y,x,str);
            curses_funcname = "mvwaddstr";
        }
        else {
            rtn = waddstr(self->win,str);
            curses_funcname = "waddstr";
        }
```

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -1064,27 +1173,37 @@ _curses_window_addnstr_impl(PyCursesWindowObject *self, int group_left_1,
+        else {
+            rtn = waddnwstr(self->win, wstr, n);
+            curses_funcname = "waddnwstr";
+        }
```
```suggestion
        simple_funcname = "addnwstr";
        if (use_xy) {
            rtn = mvwaddnwstr(self->win,y,x,wstr,n);
            curses_funcname = "mvwaddnwstr";
        }
        else {
            rtn = waddnwstr(self->win,wstr,n);
            curses_funcname = "waddnwstr";
        }
```

picnixz committed:
reduce diff x5

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -2082,27 +2234,37 @@ _curses_window_insnstr_impl(PyCursesWindowObject *self, int group_left_1,
     if (strtype == 2) {
-        funcname = "insn_wstr";
-        if (use_xy)
+        simple_funcname = "insn_wstr";
```
This one is incorrect. It should be `insnstr`.

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -892,36 +983,44 @@ _curses_window_addch_impl(PyCursesWindowObject *self, int group_left_1,
     type = PyCurses_ConvertToCchar_t(self, ch, &cch, wstr);
     if (type == 2) {
-        funcname = "add_wch";
+        simple_funcname = "add_wch";
```
Ditto. But I need to check this one actually.

picnixz commented on Modules/_cursesmodule.c:
```diff
@@ -1064,27 +1173,37 @@ _curses_window_addnstr_impl(PyCursesWindowObject *self, int group_left_1,
     if (strtype == 2) {
-        funcname = "addnwstr";
-        if (use_xy)
+        simple_funcname = "addnwstr";
```
Ditto.

picnixz reviewed (commented):
Actually, there are some `funcname` that should be corrected but I need to see how they should really be corrected.

picnixz committed:
simplify even more!

picnixz committed:
nits(cosmetics)

picnixz committed:
update messages

picnixz committed:
remove un-necessary prototypes and rearrange code

encukou commented on Modules/_cursesmodule.c:
```diff
@@ -211,6 +211,71 @@ static int curses_start_color_called = FALSE;
+#define CURSES_ERROR_VERBOSE_FORMAT         "%s() (called by %s()) returned ERR"
+#define CURSES_ERROR_NULL_FORMAT            "%s() returned NULL"
+#define CURSES_ERROR_NULL_VERBOSE_FORMAT    "%s() (called by %s()) returned NULL"
+#define CURSES_ERROR_MUST_CALL_FORMAT       "must call %s() first"
```
I don't think the names are too useful, given that most are used in the consolidated function, so they're meant be used only once.

encukou commented on Modules/_cursesmodule.c:
```diff
@@ -211,6 +211,71 @@ static int curses_start_color_called = FALSE;
+ * Format a curses error.
+ *
+ * A NULL 'python_funcname' falls back to 'curses_funcname' and vice-versa.
+ * If both names are NULL, the error message is 'catchall_ERR'.
```
Is it valid for `python_funcname` to be NUL but `curses_funcname` to be set?

I see only these cases:
- the function names are the same, or `PyCursesCheckERR` was called by the user (only `python_funcname` is non-NULL)
- they names are different (both are non-NULL)
- `PyCursesCheckERR` called by user with NULL (so both are NULL)

Is it worth an assert?

picnixz replied:
Mmh. Semantically, yes. In practice no. Sorry, it was a relic of the previous implementatiom. So either python and curses names are the different, or only python function name is given (and matches that of curses). 

I won't be able to put anything before the beta release but I don't mind delaying this one for 3.15. However if you find time, I'd be happy if you make the change!

encukou replied:
Lots to do for the beta, so 3.15 it is.

encukou commented on Modules/_cursesmodule.c:
```diff
@@ -211,6 +211,71 @@ static int curses_start_color_called = FALSE;
+}
+
+static void
+PyCursesSetError_ForWin_From(PyCursesWindowObject *win,
```
Ideally, don't use `Py`, the prefix for public API, for new static functions.

picnixz replied:
Maybe I should just have a `curses_set_error_*` prefix? I don't remember whether I unfortunately introduced this naming in this or in a previous PR :(

encukou commented on Modules/_cursesmodule.c:
```diff
@@ -3408,7 +3564,7 @@ _curses_initscr_impl(PyObject *module)
 
     if (curses_initscr_called) {
-        wrefresh(stdscr);
+        (void)wrefresh(stdscr); // TODO(picnixz): should we report an error?
```
Please open issues for TODOs.

picnixz replied:
Sorry, I'll do so when I'm back (in 10 days)

picnixz replied:
See https://github.com/python/cpython/issues/133579.

encukou replied:
```suggestion
        wrefresh(stdscr);
```

encukou reviewed (commented):
Thanks! This should add some clarity.

picnixz committed:
address review

- use `curses_set_error_*` prefix for error helpers
- remove some macros with predefined messages
- cleanly report both NULL and ERR cases

picnixz committed:
update error messages in `_curses_panel`

picnixz committed:
raise `curses.error` in `is_linetouched` instead of TypeError

picnixz committed:
Revert "update error messages in `_curses_panel`"

This reverts commit 236bea51e25b21c8f8c5b70e19221787f5200b68.

picnixz committed:
fix typo

picnixz committed:
correctly check `mouseinterval` return value

picnixz committed:
Merge remote-tracking branch 'upstream/main' into feat/curses/error-type-125843

encukou commented on Modules/_cursesmodule.c:
```diff
@@ -1286,7 +1405,7 @@ _curses_window_box_impl(PyCursesWindowObject *self, int group_right_1,
         }
     }
-    box(self->win,ch1,ch2);
+    (void)box(self->win,ch1,ch2);
```
IMO, if a compiler warns here, it does have a point.
```suggestion
    box(self->win,ch1,ch2);
```

picnixz replied:
I'll anyway change it in the next PR where I handle the error myself but I can reduce the diff in this PR.

encukou commented on Modules/_cursesmodule.c:
```diff
@@ -1384,13 +1504,15 @@ PyCursesWindow_ChgAt(PyObject *op, PyObject *args)
     if (use_xy) {
         rtn = mvwchgat(self->win,y,x,num,attr,color,NULL);
-        touchline(self->win,y,1);
+        (void)touchline(self->win,y,1);
```
```suggestion
        touchline(self->win,y,1);
```

encukou commented on Modules/_cursesmodule.c:
```diff
@@ -1384,13 +1504,15 @@ PyCursesWindow_ChgAt(PyObject *op, PyObject *args)
         getyx(self->win,y,x);
         rtn = wchgat(self->win,num,attr,color,NULL);
-        touchline(self->win,y,1);
+        (void)touchline(self->win,y,1);
```
```suggestion
        touchline(self->win,y,1);
```

encukou reviewed (approved):
OK, let's do it!
Thank you!

picnixz committed:
reduce overall diff

Co-authored-by: Petr Viktorin <encukou@gmail.com>
